### PR TITLE
fail in case nothing to run was found

### LINKED ISSUE #43121: x.py test with non-existent test directory should error
`./x.py test does-not-exist` should spit out an error instead of being successful.

cc @GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #43121.

r? @Mark-Simulacrum

### CODE REVIEW

File: src/bootstrap/step.rs [hunk 1]
```diff
@@ -1209,11 +1209,17 @@ invalid rule dependency graph detected, was a rule added and maybe typo'd?
...
                 Some((rule, 0))
             } else {
-                paths.iter().position(|path| path.ends_with(rule.path))
+                paths.iter()
+                     .position(|path| path.ends_with(rule.path))
                      .map(|priority| (rule, priority))
             }
         }).collect();
 
+        if rules.is_empty() {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/step.rs]: Hm, this is good, but I sort of expected we'd go a step further and check that every path matched with a rule... I might be wrong, but I think that should always be the case. cc @alexcrichton -- can you confirm we should match all paths?

**Comment by kennytm** [comment]
[`make prepare` failed](https://travis-ci.org/rust-lang/rust/jobs/252024851), because, [well](https://github.com/rust-lang/rust/blob/d84693b93dae3958e3504f817face0184c5c3fdd/src/bootstrap/mk/Makefile.in#L75-L76)

```make
prepare:
    $(Q)$(BOOTSTRAP) build nonexistent/path/to/trigger/cargo/metadata
```

Please either special-case "`nonexistent/path/to/trigger/cargo/metadata`" or change `make prepare` not based on a non-existing path.

File: src/bootstrap/step.rs (hunk 1 above)
Comment (alexcrichton) [inline src/bootstrap/step.rs]: Sure, seems fine.

Comment (GuillaumeGomez) [inline src/bootstrap/step.rs]: There are a lot of paths and a lot of them aren't doing anything, so I'm sure this is a good idea...

**Comment by GuillaumeGomez** [comment]
Fixed `make prepare`.

**Comment by GuillaumeGomez** [comment]
Ping in here.

**Comment by Mark-Simulacrum** [comment]
Alright, I'll accept that stopgap for now, though we should look into a better solution. It's fine for the time being I think.

@bors r+ rollup

**Comment by Mark-Simulacrum** [comment]
Bors? You okay?

@bors r-

**Pull request merged**